utils/python/CIME/hist_utils.py

```diff
@@ -0,0 +1,175 @@
+    return all_success, comments
+
+def compare_test(case, suffix1, suffix2):
+    """
+    Compares two component history files in the testcase directory
+
+    returns (SUCCESS, comments)
+    """
+    rundir   = case.get_value("RUNDIR")
+
+    return _compare_hists(case, suffix1, suffix2, rundir, rundir)
+
+def cprnc(file1, file2, case, rundir):
+    cprnc_exe = case.get_value("CCSM_CPRNC")
+    basename = os.path.basename(file1)
+    stat, out, _ = run_cmd("%s -m %s %s 2>&1 | tee %s/%s.cprnc.out" % (cprnc_exe, file1, file2, rundir, basename))
+    return (stat == 0 and "files seem to be IDENTICAL" in out, out)
+
+def compare_baseline(case, baseline_dir=None):
+    """
+    compare the current test output to a baseline result
+    """
+    rundir   = case.get_value("RUNDIR")
+    if baseline_dir is None:
+        baselineroot = case.get_value("BASELINE_ROOT")
+        basecmp_dir = os.path.join(baselineroot, case.get_value("BASECMP_CASE"))
+        dirs_to_check = (baselineroot, basecmp_dir)
+    else:
+        basecmp_dir = baseline_dir
+        dirs_to_check = (basecmp_dir,)
+
+    for bdir in dirs_to_check:
+        if not os.path.isdir(bdir):
+            return False, "ERROR %s does not exist" % bdir
+
+    return _compare_hists(case, from_dir1=rundir, from_dir2=basecmp_dir)
+
+def get_extension(model, filepath):
+    """
+    >>> get_extension("cpl", "cpl.hi.nc")
+    'hi'
+    >>> get_extension("cpl", "cpl.h.nc")
+    'h'
+    >>> get_extension("cpl", "cpl.h1.nc")
+    'h1'
+    >>> get_extension("cpl", "TESTRUNDIFF_Mmpi-serial.f19_g16_rx1.A.melvin_gnu.C.fake_testing_only_20160816_164150-20160816_164240.cpl.hi.0.nc.base")
+    'hi'
+    >>> get_extension("cpl", "TESTRUNDIFF_Mmpi-serial.f19_g16_rx1.A.melvin_gnu.C.fake_testing_only_20160816_164150-20160816_164240.cpl.h.nc")
+    'h'
+    """
+    basename = os.path.basename(filepath)
+    ext_regex = re.compile(r'.*%s.*[.](h.?)([.][^.]+)?[.]nc' % model)
+    m = ext_regex.match(basename)
+    expect(m is not None, "Failed to get extension for file '%s'" % filepath)
+    return m.groups()[0]
+
+def generate_baseline(case, baseline_dir=None):
+    """
+    copy the current test output to baseline result
+    """
+    rundir   = case.get_value("RUNDIR")
+    if baseline_dir is None:
+        baselineroot = case.get_value("BASELINE_ROOT")
+        basegen_dir = os.path.join(baselineroot, case.get_value("BASEGEN_CASE"))
+    else:
+        basegen_dir = baseline_dir
+    testcase = case.get_value("CASE")
+
+    if not os.path.isdir(basegen_dir):
+        os.makedirs(basegen_dir)
+
+    comments = "Generating baselines into '%s'\n" % basegen_dir
+    num_gen = 0
+    for model in _iter_model_file_substrs(case):
+        comments += "  generating for model '%s'\n" % model
+        hists = _get_all_hist_files(testcase, model, from_dir=rundir)
+        num_gen += len(hists)
+        for hist in hists:
+            ext = get_extension(model, hist)
+            basename = "%s.%s.nc" % (model, ext)
```

Comment:
So it looks like this generates baselines for every file, but if you have multiple files (from multiple times) with the same extension, then they just overwrite each other, so you end up with one in the end. Maybe a bit inefficient, but shouldn't be a big deal in practice.

However, this appears to be problematic in the comparison of baselines: If you have, say:

somecase.cpl.hi.0001-01-01.nc
somecase.cpl.hi.0001-01-02.nc

Then the baseline directory will just have:
cpl.hi.nc

and the comparison will fail (I think) because the number of files differs between the test case and the baseline directory.

@jedwards4b noticed something similar, and is working on a fix for this, along with possibly changing the behavior for the in-test comparisons so we just look at the last history file, as has been done in the past.